client-tags/display-name.md

```diff
@@ -0,0 +1,104 @@
+---
+title: "`display-name` client tag"
+layout: spec
+work-in-progress: true
+copyrights:
+  -
+    name: "James Wheare"
+    period: "2021"
+    email: "james@irccloud.com"
+---
+
+## Notes for implementing work-in-progress version
+
+This is a work-in-progress specification.
+
+Software implementing this work-in-progress specification MUST NOT use the
+unprefixed `+display-name` tag name. Instead, implementations SHOULD use the
+`+draft/display-name` tag name to be interoperable with other software
+implementing a compatible work-in-progress version.
+
+The final version of the specification will use an unprefixed tag name.
+
+## Introduction
+
+This specification defines a client-only message tag to indicate a display name to use for the attached message, separate from the nickname or realname of the user.
+
+## Motivation
+
+This tag provides a way to provide further temporary context about the originator of a message. Example use cases include bots that relay messages from external contexts, or in role playing environments.
+
+## Architecture
+
+For the purpose of this specification, a message is defined as a `PRIVMSG`, `NOTICE`, `TAGMSG`, or [`batch`][batch] end.
+
+### Dependencies
+
+Clients wishing to use this tag MUST negotiate the [`message-tags`][tags] capability with the server.
+
+### Format
+
+Display names are indicated by a message with a `+draft/display-name` tag using the client-only prefix with the value being the desired name.
+
+## Client implementation considerations
+
+This section is non-normative.
+
+This client tag is not meant as a method of impersonating other IRC users. Take care to make it clear to users when a display name is being used. Consider formatting display names differently to nicknames and making the original nickname visible to the user.
+
+This client tag is also not meant to be used as a persistent display name for users. The [metadata framework][metadata] would be more suitable for such a use case.
```

Comment:
Metadata mentions that it is deprecated, given that it discourages its own use and implementation it seems unfit to suggest that as more suitible for some use cases.  Would we loose anything if we dropped this last sentance out?

Reply:
Yeah it's not ideal. It will hopefully come back in future. I just want to discourage people from just sending a fixed display-name for all their own messages.

Reply:
only that specific version of the metadata spec is deprecated, a version fixing the issues that led to that deprecation is still pending

Reply:
re. my earlier comment about the name conflicting with metadata, I don't have a clear picture of how this is going to interact with future iterations of the metadata spec. If a relay bot publishes a display name for itself in metadata, but also sends this tag, is the intent that this tag should supersede the value in metadata? But if exactly one field is present, clients will treat them identically?

This capability is unprivileged and allows the client to "silently" change their display name arbitrarily many times on the same connection, which seems like it poses abuse issues that neither RELAYMSG nor metadata did. (RELAYMSG because it was a privileged operation, metadata because of the pub/sub and rate-limiting features.)

Reply:
What's the concrete abuse potential?

The intent is that the actual nick is still visible and obvious to users (same as with metadata), which is not the case with RELAYMSG. The display-name is a decoration, not a canonical replacement for the nick.

Since metadata display-name isn't yet defined, it's hard to define the relationship between a metadata and tag-based display-name, but I'd expect a per-message tag could override the metadata version for decoration of messages. That said, clients might want to show the metadata version too, perhaps in a on-click popover or some other paradigm.

I'm not sure how pub/sub or rate limiting has much to do with a client being able to choose their own unprivileged display name.

Reply:
I think there are a number of possible scenarios here, but here's one: an unprivileged user can impersonate anyone on the other side of the bridge. If mods are not present and/or the other users are not well-informed about which is the "authorized" relay bot (or if users are not paying close attention to the UI), this could be very disruptive.

Reply:
Is that not already the case though? I can change my nick to something similar to the relay bot and then just prefix my messages with `<slingamn>`

I don’t see how display names makes this any worse.

Reply:
It doesn't make it any worse; the claim is that RELAYMSG makes it better.

Reply:
It's a non-goal of this proposal to address that.